File: Algorithm/Algorithm/Algorithm/NumberOf1.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace MattFengTestMain.Algorithm
{
    /// <summary>
    /// 二进制中1的个数
    /// </summary>
    public class NumberOf1
    {
        //输入一个整数，输出该数32位二进制表示中1的个数。其中负数用补码表示。

       
        public int numberOf1(int n)
        {
            if (n == 0)
                return 0;
            else if (n > 0)
            {
                var str= Convert.ToString(n, 2);
                return Regex.Matches(str, "1").Count;
            }
            else
            {
                var str = Convert.ToString(ConvertComplementCode(n), 2);
                return Regex.Matches(str, "1").Count;
            }
        }

        /// <summary>
        /// 求一个16位数数的补码
        /// </summary>
        /// <param name="OriginalCode">传入一个Int16整型</param>
        /// <returns></returns>
        private static int ConvertComplementCode(int OriginalCode)
        {
            int a = short.MaxValue;
            int b = short.MinValue;
            int c = a - b;
            int d = 0;
            if (OriginalCode > 0)
            {
                d = -(c - OriginalCode + 1);
            }
            else
            {
                d = c + OriginalCode + 1;
            }
            return d;
        }

        /// <summary>
        /// 求一个32位数的补码
        /// </summary>
        /// <param name="OriginalCode">传入一个Int32整型</param>
        /// <returns></returns>
        private static long ConvertComplementCode(long OriginalCode)
        {
            long a = int.MaxValue;
            long b = int.MinValue;
            long c = a - b;
            long d = 0;
            if (OriginalCode > 0)
            {
                d = -(c - OriginalCode + 1);
            }
            else
            {
                d = c + OriginalCode + 1;
            }
            return d;
        }
    }
}

File: Algorithm/Algorithm_/NumberOf1.cpp
#include "NumberOf1.h"
using namespace std;
//二进制中1的个数

 /// <summary>
 /// 输入一个整数，输出该数32位二进制表示中1的个数。其中负数用补码表示。
 /// </summary>
void NumberOf1::Test()
{
		int x = 0b01; //二进制
		int y = 074; //八进制
		int s = 74; // 十进制
		int z = 0x74;//十六进制
		int ss = 0x01;
		cout << y << endl;
	 auto ret = NumberOf1_1(1);
	  ret = NumberOf1_1(10);
	  ret = NumberOf1_1(0);
	  ret = NumberOf1_1(-1);
	  ret = NumberOf1_1(-9);
}

/// <summary>
/// 考虑正负数都可以 一个个进行 & &32次；
/// </summary>
/// <param name="n"></param>
/// <returns></returns>
int   NumberOf1::NumberOf1_1(int n)
{
	int count = 0;
	int maxIndex = 1;
	while (maxIndex != 0)
	{
		if (n & maxIndex)
			count++;
		maxIndex <<= 1;

	}
	return count;
}

/// <summary>
/// 巧妙算法，n & (n - 1)  一次除掉一个 1 ；并且获取除去后的所有1； 当为负数也可以；大大减少了时间复杂度
/// </summary>
/// <param name="n"></param>
/// <returns></returns>
int   NumberOf1::NumberOf1_2(int n)
{

	int count = 0;
	while (n != 0)
	{
		count++;
		n = n & (n - 1);

	}
	return count;
}